### ISSUE TITLE
Scale to infinite number of documents.

### ISSUE THREAD

**Issue by gomain** [TheCre8tor/bim_launcher#4]
https://github.com/TheCre8tor/bim_launcher/blob/0d6c2b322b6cbe01187d41bd897b8cbda5df7e3a/src/index.ts#L21

We would have to account for a large number of documents that can't be got at once. Most api support some variant of _paging_. Say a lower-level service would be
```ts
/**
 * @param {number} perPage - number of documents in one page
 * @param {number} page - the page order to get, first page is 1
 */
async function getDocmuentPaged(perPage: number, page: number): Promise<Array<Document>> { /* ... */ }
```
How would we use this to implement an [`AsyncGenerator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator) such that we can write, in `main`,
```ts
  for await (const document of getDocuments()) {
    await saveDocument(document);
  }
```

**Comment by TheCre8tor** [comment]
According to my knowledge, the Generator function works like Streams in Dart.
Am I right?

``` Typescript
async function* generateDocuments(): AsyncGenerator<Documents> {
  // 1. So what I suggest is that we should call the getDocument() function inside the generateDocuments().
  let response = await getDocuments();

  let documents: Documents = [];
  let INDEX = 0;

  // 2. After getting a response from getDocument(), we loop through the response base on the length of the returned data.
  for (INDEX; INDEX <= response.length - 1; INDEX++) {
    // 3. We can now spread the items of the returned data into a new array, which places them in a single array.
    documents.push(...response[INDEX]);
  }

  // 4. Then we yield the result.
  yield await Promise.resolve(documents);
}
```

**Comment by TheCre8tor** [comment]
``` Typescript
async function main(): Promise<void> {
  // 1. Continously runs, make http request to an API endpoint so that it will detect chages of documents

  // 2a. call another api to write to ->
  // 2b. scalling ->
  const response = generateDocuments();
  await saveDocument((await response.next()).value);
}
```

We can now call generateDocuments() inside the main function and pass its response into the saveDocument() function.

**Comment by TheCre8tor** [comment]
Here's what I got after running the code. What do you think?
![Screenshot 2023-01-31 at 10 19 21 AM](https://user-images.githubusercontent.com/47912096/215719377-0233ffe5-9cf0-4d36-a6a4-a4cbc577f10f.png)

**Comment by gomain** [comment]
Yes, it works much like a (lazy) stream.

We want documents to be generated _as_ they are consumed; and discarded. Not upfront. So we would have to `yield` multiple times, i.e. upon the generator being called `next`.

**Comment by TheCre8tor** [comment]
Here is what I did, instead of returning a list of documents from getDocuments() method, I yield each document multiple times, and I also make use of the  `for await (...) { }` syntax instead of calling the next every time.

``` Typescript
async function main(): Promise<void> {
  // 2b. scalling ->
  for await (const document of getDocuments()) {
    const transformedDocument = transformDocument(document);
    await saveDocument(transformedDocument);
  }
}

export async function* getDocuments(): AsyncGenerator<Documents> {
  let attribute1 = new Map<string, string>();
  let attribute2 = new Map<string, string>();
  let attribute3 = new Map<string, string>();

  attribute1.set("name", "BREEZ-MJT-GH-B1-DR-A-0001.pdf");
  attribute1.set("role", "A");
  attribute1.set("version", "V1");

  attribute2.set("name", "BREEZ-MJT-GH-GF-DR-A-0001.pdf");
  attribute2.set("role", "A");
  attribute2.set("version", "V2");

  attribute3.set("name", "BREEZ-MJT-ZZ-01-DR-0002.pdf");
  attribute3.set("role", "A");
  attribute3.set("version", "V1");

  let version1: Version = attribute1;
  let version2: Version = attribute2;
  let version3: Version = attribute3;

  let document1: Documents = [version1, version2];
  let document2: Documents = [version3];

  yield document1;
  yield document2;
}
```

**Comment by gomain** [comment]
More practically, our `getDocuments` would be calling
```ts
/**
 * @param {number} perPage - number of documents in one page
 * @param {number} page - the page order to get, first page is 1
 */
async function getDocmuentPaged(perPage: number, page: number): Promise<Array<Document>> { /* ... */ }
```
under the hood.

Let's mock `getDocumentPaged` and implement `getDocuments` in terms of it. Note that `getDocumentPaged` returns an array of documents, while we want `getDocuments` to return `AsyncGenerator<Document>`.

**Comment by TheCre8tor** [comment]
Alright. Thanks

**Comment by TheCre8tor** [comment]
Hi Gomain,
Sorry I wasn't able to reply or complete the exercise yesterday,
I was very occupied. 

Here is what came to my mind while solving the problem. Since we are defining the number of documents we need per page while calling the `async function getDocmuentPaged(perPage: number, page: number): Promise<Array<Document>> { /* ... */ }`, we can loop over the response and yield each document one at a time.

``` Typescript
export async function* getDocuments(): AsyncGenerator<Documents> {
  const PER_PAGE = 2;
  const PAGE = 1;

  let response = await getDocmuentPaged(PER_PAGE, PAGE);

  for (let document of response) {
    yield document;
  }
}
```

Do you think this approach to solving the problem is good?

**Comment by TheCre8tor** [comment]
Here is another solution I came up with but I am a bit skeptical about it; 
My fear is: what if the documents return from `await getDocmuentPaged(PER_PAGE, PAGE);` is not up to the number specified in the `PER_PAGE` variable? Then it will result in an exception.

What do you think?

export async function* getDocuments(): AsyncGenerator<Documents> {
  const PER_PAGE = 2;
  const PAGE = 1;
  let INDEX = 0;

  let response = await getDocmuentPaged(PER_PAGE, PAGE);

  for (INDEX; INDEX <= PER_PAGE - 1; INDEX++) {
    yield response[INDEX];
  }
}

**Comment by gomain** [comment]
Think about it.

Hint: arrays are [Iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols)s.

**Comment by TheCre8tor** [comment]
Yeah, that was why I iterate over the response before yielding it.

**Comment by gomain** [comment]
From now on let's put changes into PRs. We move the discussions there.

**Comment by TheCre8tor** [comment]
Alright, I have pushed the latest changes I wrote to this repository's main branch.

**Comment by TheCre8tor** [comment]
> From now on let's put changes into PRs. We move the discussions there.

PRs to which repository?

**Comment by gomain** [comment]
> > From now on let's put changes into PRs. We move the discussions there.
> 
> PRs to which repository?

This one.

**Comment by TheCre8tor** [comment]
Alright, have you run the latest code I pushed to the main branch?

**Comment by gomain** [comment]
https://github.com/TheCre8tor/bim_launcher/blob/c333b6ec457590cc82ae3318b5cccf04303af1a1/src/external_service.ts#L2-L5

https://github.com/TheCre8tor/bim_launcher/blob/c333b6ec457590cc82ae3318b5cccf04303af1a1/src/external_service.ts#L19-L47

We're not getting documents of _all_ pages. `getDocumentPaged` should mock returning different documents upon being given different arguments. Imagine this actually makes an http request to some api forwarding the paging arguments.

**Comment by TheCre8tor** [comment]
Can we talk about the paging system, what make a page?

**Comment by gomain** [comment]
When there are more items of something such that it is not feasible to return all-of-them in one call. Especially over the network, such as an endpoint `GET https://some.platform.com/api/listDocumets?project=project+name`. I.e. say it would be 2GB to return all documents of the project. This is, most of the time, undesirable for both the platform and the client. Apis typically support [pagination](https://duckduckgo.com/?q=api+pagination&t=vivaldi&ia=web). Where clients can ask for a portion of the items. This is usually expressed in terms of paging as it's the typical use case, i.e. displaying lists in a web page split into pages, i.e. with next/prev buttons.

The endpoint now becomes `GET https://some.platform.com/api/listDocumets?project=project+name&per_page=250&page=3` Meaning return the 501st to 750th item, i.e. items of the 3rd page had pages contained 250 items.

**Comment by TheCre8tor** [comment]
Thanks for the explanation. I will implement it and get back to you shortly
